### PR TITLE
[WIP] Java build fixes

### PR DESCRIPTION
**Pull request by gerashegalov**
- fix warning CMAKE_CUDA_ARCHITECTURES CMP0104
- invoke make via cmake --build towards supporting other generators

Signed-off-by: Gera Shegalov <gera@apache.org>

### CODE REVIEW

File: java/pom.xml [hunk 1]
```diff
@@ -371,7 +371,6 @@
                         <phase>validate</phase>
                         <configuration>
                             <tasks>
-                                <mkdir dir="${native.build.path}"/>
```
Comment (jlowe) [inline java/pom.xml:374]: When I perform a clean build with this change in the conda environment I get:
```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-antrun-plugin:1.3:run (cmake) on project cudf: An Ant BuildException has occured: The directory /home/jlowe/src/cudf/java/target/cmake-build does not exist -> [Help 1]
```

File: java/pom.xml [hunk 2]
```diff
@@ -384,11 +383,13 @@
...
                                     <arg value="-DGPU_ARCHS=${GPU_ARCHS}"/>
+                                    <arg value="-DCMAKE_CUDA_ARCHITECTURES=${GPU_ARCHS}"/>
                                 </exec>
                                 <exec dir="${native.build.path}"
                                       failonerror="true"
-                                      executable="make">
-                                    <arg value="-j"/>
+                                      executable="cmake">
+                                    <arg line="--build ."/>
+                                    <arg line="-j ${env.PARALLEL_LEVEL}"/>
```
Comment (jlowe) [inline java/pom.xml:392]: When I built before cleaning with this in the conda env supplying cmake 3.18.4, I get:
```
     [exec] Unknown argument ${env.PARALLEL_LEVEL}
     [exec] Usage: cmake --build <dir> [options] [-- [native-options]]
```

**Review by jlowe — CHANGES_REQUESTED** [review]
This needs to target branch-0.19.

The PR headline needs to be more descriptive, as this is what will be added to the changelog.  I don't see this as fixing the build, if anything it breaks it quite badly, at least for me. :smile:

File: java/pom.xml (hunk 1 above)
Comment (gerashegalov) [reply java/pom.xml:374]: to avoid ambiguities what is the exact command you tried?

Comment (jlowe) [reply java/pom.xml:374]: Nothing fancy,  just `mvn clean test`:
```
$ mvn clean test
[INFO] Scanning for projects...
[INFO] 
[INFO] ---------------------------< ai.rapids:cudf >---------------------------
[INFO] Building cudfjni 0.19-SNAPSHOT
[INFO] --------------------------------[ jar ]---------------------------------
[INFO] 
[INFO] --- maven-clean-plugin:3.1.0:clean (default-clean) @ cudf ---
[INFO] Deleting /home/jlowe/src/nvcudf/java/target
[INFO] 
[INFO] --- maven-antrun-plugin:1.3:run (cmake) @ cudf ---
[INFO] Executing tasks
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  0.418 s
[INFO] Finished at: 2021-02-16T16:51:37-06:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-antrun-plugin:1.3:run (cmake) on project cudf: An Ant BuildException has occured: The directory /home/jlowe/src/nvcudf/java/target/cmake-build does not exist -> [Help 1]
```

Comment (gerashegalov) [reply java/pom.xml:374]: I can't repro this. What is your cmake version?

Comment (gerashegalov) [reply java/pom.xml:374]: ah reading the later comment your cmake is `3.18.4`. I tested with `3.18.5`.

File: java/pom.xml (hunk 2 above)
Comment (gerashegalov) [reply java/pom.xml:392]: If 3.18.5 fixes it maybe we can make it the minimum?

Comment (jlowe) [reply java/pom.xml:392]: Personally I don't think the Java bindings should require a cmake version beyond what is required for the rest of the cudf project.  Is there not a way to convey the build parallelism with the cmake version we already have?

Comment (gerashegalov) [reply java/pom.xml:392]: we currently have somewhat [inconsistent definitions](https://github.com/rapidsai/cudf/search?q=cmake_minimum_required) `cmake_minimum_required` in the repo.

it's not just about specifying the parallelism but also the ability to use different generators later on.

Comment (jlowe) [reply java/pom.xml:392]: > also the ability to use different generators later on.

I don't see this as a pressing need to warrant requiring a cmake version beyond anything else currently in the project.

Comment (jlowe) [reply java/pom.xml:392]: On a related note, it looks like the project will be moving to 3.18 shortly.  See https://github.com/rapidsai/cudf/pull/7107#discussion_r579376674.

**Comment by gerashegalov** [comment]
closing for now

**Pull request closed without merging**